sdk/python/kfp/components/_components.py

```diff
@@ -537,7 +537,12 @@ def resolve_argument(argument):
 
         task_obj = task_factory(**pythonic_task_arguments)
         task_outputs_with_pythonic_names = task_obj.outputs
-        task_outputs_with_original_names = {pythonic_output_name_to_original[pythonic_output_name]: output_value for pythonic_output_name, output_value in task_outputs_with_pythonic_names.items()}
+        task_outputs_with_original_names = {
+            # component_bridge generates outputs under both pythonic and original name,
+            # so half of them are absent from pythonic_output_name_to_original
+            pythonic_output_name_to_original.get(pythonic_output_name, pythonic_output_name): output_value
+            for pythonic_output_name, output_value in task_outputs_with_pythonic_names.items()
+        }
```

Comment:
I think this can be simpler:
```suggestion
        task_outputs_with_original_names = {output.name: task_obj.outputs[output.name] for output in task_component_spec.outputs or []}
```

`output_name_to_pythonic`, `pythonic_output_name_to_original` and `task_outputs_with_pythonic_names ` are not needed then.

Reply:
@Ark-kun
Was it ever a part of the contract for `_container_task_constructor`? It's not backwards-compatible, but it certainly would simplify things.

I wasn't sure about it, considering that `_container_task_constructor` can be overwritten and I only know of two overwrites:
* `_components._create_task_spec_from_component_and_arguments` --- which [uses](../blob/b1f46accdfeef67b2ff31e16e10304a6412f6b14/sdk/python/kfp/components/_components.py#L239) original outputs' names, as seen in [`TaskSpec._init_outputs`](../blob/b1f46accdfeef67b2ff31e16e10304a6412f6b14/sdk/python/kfp/components/_structures.py#L580-L596)
* `_component_bridghe._create_container_op_from_component_and_arguments` --- which uses both, as shown in the issue

Reply:
Would the following be ok?

```python
KT = TypeVar('KT') # could be imported from typing, of course
VT = TypeVar('VT')
def get_maybe_remap(m: Mapping[KT, VT], key: KT,
                    key_mapper: Callable[[KT], KT]) -> VT:
    if key in m:
        return m[key]
    else:
        return m[key_mapper(key)]
(...)

task_outputs_with_original_names = {
    output.name: get_maybe_pythonic(task_obj.outputs, output.name,
                                    output_name_to_pythonic.__getitem__)
    for output in task_component_spec.outputs or []
}

```

It would remove the need for `pythonic_output_name_to_original` (i.e. reverse-mapping of `output_name_to_pythonic`) and `task_outputs_with_pythonic_names`, leaving just `output_name_to_pythonic` "just in case", for backward-compatibility. It's also reusable.

Reply:
> Was it ever a part of the contract for `_container_task_constructor`? It's not backwards-compatible, but it certainly would simplify things.

I think this should be the contract (original output names should be the keys in the `task.outputs` dictionary).
I think at his point it's still OK it make this change. The contract is private and I do not think anyone else has implemented the contact yet (also we're making a significantly more drastic change to it very soon).
We can always restore the behavior is we encounter unfixable case in the future.

Reply:
@Ark-kun Ok, if it can be changed, then I'll simplify it and add the explicit contract.